src/NHibernate/Proxy/FieldInterceptorProxyBuilder.cs

```diff
@@ -0,0 +1,298 @@
+		private static readonly MethodInfo FieldInterceptorMarkDirtyMethod = FieldInterceptorType.GetMethod(nameof(IFieldInterceptor.MarkDirty));
+		private static readonly System.Type AbstractFieldInterceptorType = typeof(AbstractFieldInterceptor);
+		private static readonly FieldInfo AbstractFieldInterceptorInvokeImplementationField =
+			AbstractFieldInterceptorType.GetField(nameof(AbstractFieldInterceptor.InvokeImplementation));
+
+		public static TypeInfo CreateProxyType(System.Type baseType)
+		{
+			// Avoid having a suffix ending with "Proxy", for disambiguation with INHibernateProxy proxies
+			var typeName = $"{baseType.Name}ProxyForFieldInterceptor";
+			var assemblyName = $"{typeName}Assembly";
+			var moduleName = $"{typeName}Module";
+
+			var name = new AssemblyName(assemblyName);
+
+			var assemblyBuilder = ProxyBuilderHelper.DefineDynamicAssembly(AppDomain.CurrentDomain, name);
+			var moduleBuilder = ProxyBuilderHelper.DefineDynamicModule(assemblyBuilder, moduleName);
+
+			const TypeAttributes typeAttributes = TypeAttributes.AutoClass | TypeAttributes.Class | TypeAttributes.Public | TypeAttributes.BeforeFieldInit;
+
+			var interfaces = new HashSet<System.Type>
+			{
+				typeof(IFieldInterceptorAccessor)
+			};
+
+			// Use the object as the base type
+			// since we're not inheriting from any class type
+			var parentType = baseType;
+			if (baseType.IsInterface)
+			{
+				parentType = typeof(object);
+				interfaces.Add(baseType);
+			}
+
+			interfaces.RemoveWhere(i => !i.IsVisible);
+
+			var typeBuilder = moduleBuilder.DefineType(typeName, typeAttributes, parentType, interfaces.ToArray());
+
+			var fieldInterceptorField = typeBuilder.DefineField("__fieldInterceptor", FieldInterceptorType, FieldAttributes.Private);
+
+			ImplementConstructor(typeBuilder, parentType);
+
+			foreach (var method in ProxyBuilderHelper.GetProxiableMethods(baseType, interfaces))
+			{
+				CreateProxiedMethod(typeBuilder, method, fieldInterceptorField);
+			}
+
+			// Make the proxy serializable
+			var serializableConstructor = typeof(SerializableAttribute).GetConstructor(System.Type.EmptyTypes);
+			var customAttributeBuilder = new CustomAttributeBuilder(serializableConstructor, Array.Empty<object>());
+			typeBuilder.SetCustomAttribute(customAttributeBuilder);
+			// Default serialization is enough to handle serialization/de-serialization in the same process.
+			// For inter-process serialization, the proxy factory should be serialized too, for ensuring
+			// re-creation of the proxy type. The static NHibernateProxy does only that, letting the entity state
+			// be restored by lazy load. The field interceptor proxy cannot do that: it only knows how to load
+			// lazy properties, and not the other ones. (Moreover, its session is not restored on persistence
+			// context deserialization, contrary to NHibernateProxy. This causes deserialized proxies to only be
+			// usable for accessing data already loaded before serialization.)
```

Comment:
Those limitations are the same than those of dynamic proxies serialization of a field interceptor proxy.